### ISSUE TITLE
Introduce a `mark_sharding` that also shards the backward

### ISSUE THREAD

**Issue by tengyifei** [pytorch/xla#8678]
## 🚀 Feature

PyTorch/XLA `xs.mark_sharding` is an in-place operation that adds sharding annotation to an XLA tensor. However, gradients to be applied to the tensor are not annotated with sharding annotations.

## Motivation

In some cases, GSPMD fails to propagate sharding annotation from the tensor to its gradient. It's useful to shard both tensor and its gradient with the same sharding annotation.

## Pitch

We could write a `torch.autograd.Function` implementation to do this.

## Additional context

JAX `mark_sharding` shards the gradients too.

cc @bhavya01

**Comment by miladm** [comment]
cc @lsy323 for viz as he is working on `disabling_spmd()` API. It seems the two solutions are looking at a new subset of live tensors.

**Comment by bhavya01** [comment]
Also going to add `xs.mark_sharding_and_gradients` before closing this issue.

**Comment by rpsilva-aws** [comment]
Can you explain in which cases the sharding is not present for the gradients? We mentioned "in some cases", but does that point to a particular issue? I have only ever seen this happening in very particular cases, e.g., we pre-initialized gradients and then re-assign it to the parameter (after computing the gradients); we also saw this when adding backward hooks for SP, where we needed to explicitly shard the gradients. In which scenario would we not want to propagate the sharding to the gradients?

**Comment by tengyifei** [comment]
>  In which scenario would we not want to propagate the sharding to the gradients?

I cannot think of a single scenario where we don't want the same sharding on the gradients. In fact `jax.lax.with_sharding_constraints` just does this by default.

**Comment by rpsilva-aws** [comment]
Odd. I was hoping that this would be the assumption behind `mark_sharding` too, but I could also see the concern of changing the existing behavior if there are actual legitimate cases for not wanting it on the gradients (though I've only ever seen it in those 2 specific cases). It would be nice to eventually converge the UX and the behavior, and maybe present users with a parameter in `mark_sharding`/flag/other that does otherwise (not propagate it).

**Comment by tengyifei** [comment]
Looks like the remaining item is to add a wrapper over the autograd Function introduced in #8723 , and unit test it

**Comment by tengyifei** [comment]
@pgmoka if you would like to work on this you can change the assignee I think

**Comment by iwknow** [comment]
> Looks like the remaining item is to add a wrapper over the autograd Function introduced in [#8723](https://github.com/pytorch/xla/pull/8723) , and unit test it

i wonder what wrapper you need here? `xs.mark_sharding_and_gradients` mentioned above?
I am interested in contributing to this project and this issue is the latest issue tagged as `good first issue`.

**Comment by tengyifei** [comment]
> i wonder what wrapper you need here? xs.mark_sharding_and_gradients mentioned above?

yes, that and adding a unit test in python. we need to verify that if `t` is sharded, then `t.grad` is also sharded similarly after calling a `loss.backward()`

**Comment by iwknow** [comment]
> > i wonder what wrapper you need here? xs.mark_sharding_and_gradients mentioned above?
> 
> yes, that and adding a unit test in python. we need to verify that if `t` is sharded, then `t.grad` is also sharded similarly after calling a `loss.backward()`
sounds good. i will take a look in the next few days.

**Comment by iwknow** [comment]
i think the [initial](https://github.com/pytorch/xla/pull/8723/commits/061bb8310c77ec267b4ef8ac242a42f39cd59d1e) already the functionality we want. where do you want me to export this "MarkShardingFunction"? do you have an example of such a wrapper? thanks!

**Comment by tengyifei** [comment]
@iwknow let's add a `mark_sharding_and_gradients` function after the `mark_sharding` function at https://github.com/pytorch/xla/blob/1db6c861e393b780777bf98c55941263dc6d0315/torch_xla/distributed/spmd/xla_sharding.py#L514.

In the documentation of the `mark_sharding_and_gradients` function, it could simply say that it's a version of `mark_sharding` that will shard the gradients with the same partition spec.

**Comment by iwknow** [comment]
the development is blocked by https://github.com/pytorch/xla/issues/8813

**Comment by tengyifei** [comment]
Fixed by https://github.com/pytorch/xla/pull/8826
